File: frontend/src/types/event.ts
export const EVENT_STATUSES = ["Upcoming", "Confirmed", "Completed", "Cancelled"] as const;

export type EventStatus = (typeof EVENT_STATUSES)[number];

export const EVENT_STATUS_LABELS: Record<EventStatus, string> = {
  Upcoming: "Upcoming",
  Confirmed: "Confirmed",
  Completed: "Completed",
  Cancelled: "Cancelled",
};

export interface StudioEvent {
  eventId: number;
  customerId: number;
  customerName: string;
  customerMobileNumber: string;
  eventTypeId: number | null;
  eventTypeName: string | null;
  eventDate: string;
  startTime: string | null;
  endTime: string | null;
  venue: string | null;
  venueAddress: string | null;
  budget: number | null;
  amountPaid: number;
  balance: number;
  eventStatus: EventStatus;
  fileLocation: string | null;
  notes: string | null;
  // Stamped the first time the event is marked Completed; null for events completed before this
  // was recorded.
  completedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreateEventRequest {
  customerId: number;
  eventTypeId?: number;
  eventDate: string;
  startTime?: string;
  endTime?: string;
  venue?: string;
  venueAddress?: string;
  budget?: number;
  // Only honoured when creating: recorded as a Completed payment against the new event.
  advancePaid?: number;
  advancePaymentMethod?: string;
  eventStatus: EventStatus;
  fileLocation?: string;
  notes?: string;
}

export type UpdateEventRequest = CreateEventRequest;

// ---- Event history: the permanent record of one event ----

export interface EventHistoryQuotation {
  quotationId: number;
  // "V1", "V2", ... in the order they were raised for this event.
  version: string;
  quotationNumber: string;
  quotationDate: string;
  status: string;
  grandTotal: number;
  isApproved: boolean;
}

export interface EventHistoryPayment {
  paymentId: number;
  paymentDate: string;
  amount: number;
  paymentMethod: string;
  paymentStatus: string;
  referenceNumber: string | null;
  notes: string | null;
}

export interface EventHistoryWorker {
  workerId: number;
  workerName: string;
  workerTypeName: string | null;
  role: string | null;
}

export interface EventHistory {
  event: StudioEvent;
  workers: EventHistoryWorker[];
  quotations: EventHistoryQuotation[];
  approvedQuotation: EventHistoryQuotation | null;
  payments: EventHistoryPayment[];
  finalAmount: number | null;
}

File: backend/StudioManagement.Data/Entities/Event.cs
namespace StudioManagement.Data.Entities;

public class Event : ITenantEntity
{
    public int EventId { get; set; }
    public int StudioId { get; set; }
    public int CustomerId { get; set; }
    public int? EventTypeId { get; set; }
    public DateTime EventDate { get; set; }
    public TimeSpan? StartTime { get; set; }
    public TimeSpan? EndTime { get; set; }
    public string? Venue { get; set; }
    public string? VenueAddress { get; set; }
    public decimal? Budget { get; set; }
    public string EventStatus { get; set; } = null!;
    public string? FileLocation { get; set; }
    public string? Notes { get; set; }

    // Set the moment the event is marked Completed, so the historical record keeps the real
    // completion date even if the row is touched again later.
    public DateTime? CompletedAt { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    public Studio Studio { get; set; } = null!;
    public Customer Customer { get; set; } = null!;
    public EventType? EventType { get; set; }
    public ICollection<EventWorker> EventWorkers { get; set; } = new List<EventWorker>();
    public ICollection<Quotation> Quotations { get; set; } = new List<Quotation>();
    public ICollection<Payment> Payments { get; set; } = new List<Payment>();
    public ICollection<Expense> Expenses { get; set; } = new List<Expense>();
}
